File: software/examples/vbnet/ExampleScanForTags.vb
Imports System
Imports Tinkerforge

Module ExampleScanForTags
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your NFC Bricklet

    ' Callback subroutine for reader state changed callback
    Sub ReaderStateChangedCB(ByVal sender As BrickletNFC, ByVal state As Byte, _
                             ByVal idle As Boolean)
        If state = BrickletNFC.READER_STATE_REQUEST_TAG_ID_READY Then
            Dim i As Integer
            Dim tagType As Byte
            Dim tagID() As Byte
            Dim tagInfo As String

            sender.ReaderGetTagID(tagType, tagID)

            tagInfo = String.Format("Found tag of type {0} with ID [", tagType)

            For i = 0 To tagID.Length - 1
                tagInfo += String.Format("0x{0:02X}", tagID(i))

                If i < tagID.Length - 1 Then
                    tagInfo += " "
                End If
            Next

            tagInfo += "]"

            Console.WriteLine(tagInfo)
        End If

        If idle Then
            sender.ReaderRequestTagID()
        End If
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim nfc As New BrickletNFC(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register reader state changed callback to subroutine ReaderStateChangedCB
        AddHandler nfc.ReaderStateChangedCallback, AddressOf ReaderStateChangedCB

        ' Enable reader mode
        nfc.SetMode(BrickletNFC.MODE_READER)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module

File: software/examples/delphi/ExampleScanForTags.pas
program ExampleScanForTags;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletNFC;

type
  TExample = class
  private
    ipcon: TIPConnection;
    nfc: TBrickletNFC;
  public
    procedure ReaderStateChangedCB(sender: TBrickletNFC; const state: byte;
                                   const idle: boolean);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change XYZ to the UID of your NFC Bricklet }

var
  e: TExample;

{ Callback procedure for reader state changed callback }
procedure TExample.ReaderStateChangedCB(sender: TBrickletNFC; const state: byte;
                                        const idle: boolean);
var i: byte; var tagType: byte; var tagInfo: string; var tagID: TArrayOfUInt8;
begin
  if (state = BRICKLET_NFC_READER_STATE_REQUEST_TAG_ID_READY) then begin
    sender.ReaderGetTagID(tagType, tagID);

    tagInfo := 'Found tag of type ' + IntToStr(tagType) + ' with ID [';

    for i := 0 to (Length(tagID) - 1) do begin
      tagInfo := tagInfo + '0x' + IntToHex(tagID[i], 2);

      if i < Length(tagID) - 1 then begin
        tagInfo := tagInfo + ' ';
      end;
    end;

    tagInfo := tagInfo + ']';

    WriteLn(tagInfo);
  end;
  if (idle) then begin
    sender.ReaderRequestTagID;
  end;
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  nfc := TBrickletNFC.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register reader state changed callback to procedure ReaderStateChangedCB }
  nfc.OnReaderStateChanged := {$ifdef FPC}@{$endif}ReaderStateChangedCB;

  { Enable reader mode }
  nfc.SetMode(BRICKLET_NFC_MODE_READER);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.
